pytket/phir/sharding/sharder.py

```diff
@@ -170,16 +145,93 @@ def _build_shard(self, command: Command) -> None:
             bits_read,
             depends_upon,
         )
+
+        self._mark_dependencies(shard)
+
         self._shards.append(shard)
         logger.debug("Appended shard: %s", shard)
 
+    def _resolve_shard_dependencies(
+        self, qubits: set[Qubit], bits_written: set[Bit], bits_read: set[Bit]
+    ) -> set[int]:
+        """Finds the dependent shards for a given shard.
+
+        This involves checking for qubit interaction and classical hazards of
+        various types.
+
+        Args:
+            shard: Shard to run dependency calculation on
+            qubits: Set of all qubits interacted with in the command/sub-commands
+            bits_written: Classical bits the command/sub-commands write to
+            bits_read: Classical bits the command/sub-commands read from
+        """
+        logger.debug(
+            "Resolving shard dependencies with qubits=%s bits_written=%s bits_read=%s",
+            qubits,
+            bits_written,
+            bits_read,
+        )
+
+        depends_upon: set[int] = set()
+
+        for qubit in qubits:
```

Comment:
These iterations should be extremely fast as they only have to iterate the bits/qubits of the command in question. Could be optimized slightly further but I am almost certain not worth the effort versus the explicitness here.